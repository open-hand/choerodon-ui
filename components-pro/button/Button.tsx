import React, { Children } from 'react';
import PropTypes from 'prop-types';
import { Cancelable, DebounceSettings } from 'lodash';
import omit from 'lodash/omit';
import debounce from 'lodash/debounce';
import isString from 'lodash/isString';
import { computed, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import isPromise from 'is-promise';
import { ProgressType } from 'choerodon-ui/lib/progress/enum';
import { getConfig } from 'choerodon-ui/lib/configure';
import Icon from '../icon';
import FormContext from '../form/FormContext';
import Progress from '../progress';
import Ripple from '../ripple';
import { ButtonColor, ButtonType, FuncType } from './enum';
import { DataSetStatus } from '../data-set/enum';
import { Size, WaitType } from '../core/enum';
import DataSetComponent, { DataSetComponentProps } from '../data-set/DataSetComponent';
import autobind from '../_util/autobind';

export interface ButtonProps extends DataSetComponentProps {
  /**
   * 按钮展现形式
   * @default 'raised'
   */
  funcType?: FuncType;
  /**
   * 按钮颜色风格
   * @default 'default'
   */
  color?: ButtonColor;
  /**
   * 按钮是否是加载状态
   */
  loading?: boolean;
  /**
   * 按钮图标
   */
  icon?: string;
  /**
   * 点击跳转的地址，指定此属性 button 的行为和 a 链接一致
   */
  href?: string;
  /**
   * 相当于 a 链接的 target 属性，href 存在时生效
   */
  target?: string;
  /**
   * 点击间隔时间
   */
  wait?: number;
  /**
   * 点击间隔类型，可选值：throttle | debounce
   * @default throttle
   */
  waitType?: WaitType;
  /**
   * 按钮类型
   * @default 'button'
   */
  type?: ButtonType;
  name?: string;
  value?: any;
  form?: string;
  formAction?: string;
  formEncType?: string;
  formMethod?: string;
  formNoValidate?: boolean;
  formTarget?: string;
}

@observer
export default class Button extends DataSetComponent<ButtonProps> {
  static displayName = 'Button';

  // eslint-disable-next-line camelcase
  static __PRO_BUTTON = true;

  static contextType = FormContext;

  static propTypes = {
    /**
     * 按钮展现模式
     * 可选值：'flat' | 'raised'
     * @default raised
     */
    funcType: PropTypes.oneOf([FuncType.flat, FuncType.raised]),
    /**
     * 按钮颜色风格
     * 可选值：'default' | 'primary' | 'gray' | 'blue' | 'red' | 'green' | 'yellow' | 'purple' | 'dark'
     * @default 'default'
     */
    color: PropTypes.oneOf([
      ButtonColor.default,
      ButtonColor.primary,
      ButtonColor.gray,
      ButtonColor.blue,
      ButtonColor.red,
      ButtonColor.green,
      ButtonColor.yellow,
      ButtonColor.purple,
      ButtonColor.dark,
    ]),
    /**
     * 按钮类型
     * 可选值：'button' | 'submit' | 'reset'
     * @default 'button'
     */
    type: PropTypes.oneOf([ButtonType.button, ButtonType.submit, ButtonType.reset]),
    /**
     * 按钮是否是加载状态
     */
    loading: PropTypes.bool,
    /**
     * 点击跳转的地址，指定此属性 button 的行为和 a 链接一致
     */
    href: PropTypes.string,
    /**
     * 相当于 a 链接的 target 属性，href 存在时生效
     */
    target: PropTypes.string,
    /**
     * 点击等待时间
     */
    wait: PropTypes.number,
    /**
     * 点击间隔类型，可选值：throttle | debounce
     * @default throttle
     */
    waitType: PropTypes.oneOf([WaitType.throttle, WaitType.debounce]),
    ...DataSetComponent.propTypes,
  };

  static defaultProps = {
    suffixCls: 'btn',
    type: ButtonType.button,
    waitType: WaitType.throttle,
  };

  @computed
  get loading(): boolean {
    const { type, dataSet, loading } = this.observableProps;
    return (
      loading ||
      (type === ButtonType.submit && !!dataSet && dataSet.status === DataSetStatus.submitting)
    );
  }

  set loading(loading: boolean) {
    runInAction(() => {
      this.observableProps.loading = loading;
    });
  }

  handleClickWait: Function & Cancelable;

  constructor(props, context) {
    super(props, context);
    this.handleClickWait = this.getHandleClick(props);
  }

  getObservableProps(props, context) {
    return {
      ...super.getObservableProps(props, context),
      dataSet: 'dataSet' in props ? props.dataSet : context.dataSet,
      loading: 'loading' in props ? props.loading : this.observableProps ? this.loading : false,
      type: props.type,
      disabled: context.disabled || props.disabled,
    };
  }

  componentWillReceiveProps(nextProps, nextContext) {
    super.componentWillReceiveProps(nextProps, nextContext);
    const { wait, waitType } = this.props;
    if (wait !== nextProps.wait || waitType !== nextProps.waitType) {
      this.handleClickWait = this.getHandleClick(nextProps);
    }
  }

  componentWillUnmount() {
    this.handleClickWait.cancel();
  }

  getHandleClick(props): Function & Cancelable {
    const { wait, waitType } = props;
    if (wait && waitType) {
      const options: DebounceSettings = { leading: true, trailing: true };
      if (waitType === WaitType.throttle) {
        options.trailing = false;
        options.maxWait = wait;
      } else if (waitType === WaitType.debounce) {
        options.leading = false;
      }
      return debounce(this.handleClick, wait, options);
    }
    return debounce(this.handleClick, 0);
  }

  @autobind
  handleClickIfBubble(e) {
    const { wait, waitType } = this.props;
    if (wait && waitType) {
      e.stopPropagation();
      this.handleClickWait(e);
    } else {
      this.handleClick(e);
    }
  }

  @autobind
  async handleClick(e) {
    const { onClick } = this.props;
    if (onClick) {
      const afterClick: any = onClick(e);
      if (isPromise(afterClick)) {
        try {
          this.loading = true;
          await afterClick;
        } finally {
          this.loading = false;
        }
      }
    }
  }

  isDisabled(): boolean {
    return super.isDisabled() || this.loading;
  }

  getOmitPropsKeys(): string[] {
    return super.getOmitPropsKeys().concat([
      'icon',
      'funcType',
      'color',
      'loading',
      'wait',
      'waitType',
    ]);
  }

  getOtherProps() {
    const otherProps = super.getOtherProps();
    if (!this.disabled) {
      otherProps.onClick = this.handleClickIfBubble;
    }
    return otherProps;
  }

  getClassName(...props): string | undefined {
    const {
      prefixCls,
      props: {
        color = getConfig('buttonColor'),
        funcType = getConfig('buttonFuncType'),
        children,
        icon,
      },
    } = this;
    const childrenCount = Children.count(children);
    return super.getClassName(
      {
        [`${prefixCls}-${funcType}`]: funcType,
        [`${prefixCls}-${color}`]: color,
        [`${prefixCls}-icon-only`]: icon
          ? childrenCount === 0 || children === false
          : childrenCount === 1 && (children as any).type && (children as any).type.__C7N_ICON,
      },
      ...props,
    );
  }

  render() {
    const { children, icon, href } = this.props;
    const buttonIcon: any = this.loading ? (
      <Progress key="loading" type={ProgressType.loading} size={Size.small} />
    ) : (
      icon && <Icon type={icon} />
    );
    const hasString = Children.toArray(children).some(child => isString(child));
    const Cmp = href ? 'a' : 'button';
    const props = this.getMergedProps();
    return (
      <Ripple disabled={this.disabled}>
        <Cmp {...(href ? omit(props, ['type']) : props)}>
          {buttonIcon}
          {hasString ? <span>{children}</span> : children}
        </Cmp>
      </Ripple>
    );
  }
}
