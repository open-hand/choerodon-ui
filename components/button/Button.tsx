import React, { Children, Component, CSSProperties, FormEventHandler, KeyboardEventHandler, ButtonHTMLAttributes, MouseEventHandler } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import omit from 'lodash/omit';
import Icon from '../icon';
import Group from './ButtonGroup';
import Ripple from '../ripple';
import { Size } from '../_util/enum';
import { getPrefixCls } from '../configure';

export type ButtonType = 'primary' | 'ghost' | 'dashed' | 'danger';
export type ButtonShape = 'circle' | 'circle-outline';
export type ButtonFuncType = 'raised' | 'flat';

export interface ButtonProps {
  type?: ButtonType;
  htmlType?: string;
  icon?: string;
  shape?: ButtonShape;
  size?: Size;
  onClick?: FormEventHandler<any>;
  onMouseUp?: FormEventHandler<any>;
  onMouseDown?: FormEventHandler<any>;
  onKeyPress?: KeyboardEventHandler<any>;
  onKeyDown?: KeyboardEventHandler<any>;
  tabIndex?: number;
  loading?: boolean | { delay?: number };
  disabled?: boolean;
  style?: CSSProperties;
  prefixCls?: string;
  className?: string;
  ghost?: boolean;
  target?: string;
  href?: string;
  download?: string;
  funcType?: ButtonFuncType;
}

export default class Button extends Component<ButtonProps, any> {
  static displayName = 'Button';
  static Group: typeof Group;
  static __ANT_BUTTON = true;

  static defaultProps = {
    loading: false,
    ghost: false,
    funcType: 'flat',
  };

  static propTypes = {
    type: PropTypes.string,
    shape: PropTypes.oneOf(['circle', 'circle-outline']),
    size: PropTypes.oneOf([Size.large, Size.default, Size.small]),
    htmlType: PropTypes.oneOf(['submit', 'button', 'reset']),
    onClick: PropTypes.func,
    loading: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    className: PropTypes.string,
    icon: PropTypes.string,
    funcType: PropTypes.oneOf(['raised', 'flat']),
  };

  timeout: number;
  delayTimeout: number;

  constructor(props: ButtonProps) {
    super(props);
    this.state = {
      loading: props.loading,
      clicked: false,
    };
  }

  componentWillReceiveProps(nextProps: ButtonProps) {
    const currentLoading = this.props.loading;
    const loading = nextProps.loading;

    if (currentLoading) {
      clearTimeout(this.delayTimeout);
    }

    if (typeof loading !== 'boolean' && loading && loading.delay) {
      this.delayTimeout = window.setTimeout(() => this.setState({ loading }), loading.delay);
    } else {
      this.setState({ loading });
    }
  }

  componentWillUnmount() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
    if (this.delayTimeout) {
      clearTimeout(this.delayTimeout);
    }
  }

  handleClick: MouseEventHandler<HTMLButtonElement | HTMLAnchorElement> = (e) => {
    clearTimeout(this.timeout);
    this.timeout = window.setTimeout(() => this.setState({ clicked: false }), 500);

    const onClick = this.props.onClick;
    if (onClick) {
      onClick(e);
    }
  };

  render() {
    const {
      prefixCls: customizePrefixCls,
      type,
      shape,
      size,
      className,
      htmlType,
      children,
      icon,
      ghost,
      funcType,
      ...others
    } = this.props;

    const { loading, clicked } = this.state;

    const prefixCls = getPrefixCls('btn', customizePrefixCls);

    // large => lg
    // small => sm
    let sizeCls = '';
    switch (size) {
      case Size.large:
        sizeCls = 'lg';
        break;
      case Size.small:
        sizeCls = 'sm';
      default:
        break;
    }

    const ComponentProp = others.href ? 'a' : 'button';

    const classes = classNames(prefixCls, className, {
      [`${prefixCls}-${type}`]: type,
      [`${prefixCls}-${shape}`]: shape,
      [`${prefixCls}-${sizeCls}`]: sizeCls,
      [`${prefixCls}-icon-only`]: !children && icon,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-clicked`]: clicked,
      [`${prefixCls}-background-ghost`]: ghost,
      [`${prefixCls}-${funcType}`]: funcType,
    });

    let iconNode: any = icon ? <Icon type={icon} /> : null;
    iconNode = loading ? (
      <div className="btn-loading">
        <span className="dot1" />
        <span className="dot2" />
        <span className="dot3" />
      </div>
    ) : iconNode;
    const kids = (children || children === 0)
      ? Children.map(children, child => {
        if (typeof child === 'string') {
          return <span>{child}</span>;
        }
        return child;
      }) : null;

    return (
      <Ripple disabled={others.disabled}>
        <ComponentProp
          {...omit(others, ['loading'])}
          // 如果没有href属性，则表示组件使用button标签，type为'submit' | 'reset' | 'button'
          type={others.href ? undefined : (htmlType as ButtonHTMLAttributes<any>['type'] || 'button')}
          className={classes}
          onClick={this.handleClick}
        >
          {iconNode}
          {kids}
        </ComponentProp>
      </Ripple>
    );
  }
}
