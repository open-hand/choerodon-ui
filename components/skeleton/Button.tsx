import React, { FunctionComponent, memo, useContext } from 'react';
import omit from 'lodash/omit';
import classNames from 'classnames';
import Element, { SkeletonElementProps } from './Element';
import ConfigContext from '../config-provider/ConfigContext';

export interface SkeletonButtonProps extends Omit<SkeletonElementProps, 'size'> {
  size?: 'large' | 'small' | 'default';
}

const SkeletonButton: FunctionComponent<SkeletonButtonProps> = function SkeletonButton(props) {
  const { prefixCls: customizePrefixCls, className, active } = props;
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('skeleton', customizePrefixCls);
  const cls = classNames(prefixCls, className, `${prefixCls}-element`, {
    [`${prefixCls}-active`]: active,
  });
  return (
    <div className={cls}>
      <Element prefixCls={`${prefixCls}-button`} {...omit(props, ['prefixCls'])} />
    </div>
  );
};

SkeletonButton.displayName = 'SkeletonButton';

const MemoSkeletonButton: typeof SkeletonButton = memo(SkeletonButton);

MemoSkeletonButton.defaultProps = {
  size: 'default',
};

export default MemoSkeletonButton;
