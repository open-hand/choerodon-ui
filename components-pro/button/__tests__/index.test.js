import React from 'react';
import { render, mount } from 'enzyme';
import Button from '..';
import Icon from '../../icon';

describe('Button', () => {
  it('renders components-pro correctly', () => {
    const wrapper = render(<Button> Button-Pro </Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('have static property for components-pro type detecting', () => {
    const wrapper = mount(<Button>Button-Pro Text</Button>);
    // eslint-disable-next-line
    expect(wrapper.type().__PRO_BUTTON).toBe(true);
  });

  it('renders Chinese characters correctly', () => {
    const wrapper = render(<Button>按钮</Button>);
    expect(wrapper).toMatchSnapshot();
    // should not insert space when there is icon
    const wrapper1 = render(<Button icon="search">按钮</Button>);
    expect(wrapper1).toMatchSnapshot();
    // should not insert space when there is icon
    const wrapper2 = render(
      <Button>
        <Icon type="search" />
        按钮
      </Button>,
    );
    expect(wrapper2).toMatchSnapshot();
  });

  it('renders color correctly', () => {
    const wrapper = mount(<Button>按钮</Button>);
    expect(wrapper.props().color).toEqual(undefined);
    expect(wrapper.find('button.c7n-pro-btn-default')).toHaveLength(1);
    
    wrapper.setProps({ color: 'default', children: 'default' });
    wrapper.update();
    expect(wrapper.props().color).toEqual('default');
    expect(wrapper.find('button.c7n-pro-btn-default')).toHaveLength(1);

    wrapper.setProps({ color: 'primary', children: 'primary' });
    wrapper.update();
    expect(wrapper.props().color).toEqual('primary');
    expect(wrapper.find('button.c7n-pro-btn-primary')).toHaveLength(1);
  });

  it('renders funcType of button color and background correctly', () => {
    const wrapper = mount(<Button color="red">Red</Button>);
    expect(wrapper.props().color).toEqual('red');

    const wrapper1 = mount(
      <Button funcType="flat" color="primary">
        Blue
      </Button>,
    );
    expect(wrapper1.props().funcType).toEqual('flat');
  });

  it('button-pro should support link button', () => {
    const wrapper = mount(
      <Button target="_blank" href="https://choerodon.github.io/choerodon-ui/">
        button-pro link
      </Button>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  describe('Button funcType', () => {
    it('funType defalut value is raised', () => {
      const wrapper = mount(<Button>Raised</Button>);

      expect(wrapper.find('.c7n-pro-btn-raised').length).toBe(1);
    });

    it('set funType value is flat', () => {
      const wrapper = mount(<Button funcType='flat'>flat</Button>);

      expect(wrapper.find('.c7n-pro-btn-flat').length).toBe(1);
    });

    it('set funType value is link', () => {
      const wrapper = mount(<Button funcType='link'>link</Button>);

      expect(wrapper.find('.c7n-pro-btn-link').length).toBe(1);
    });
  })

  describe('Button loading', () => {
    it('loading defalut value is false', () => {
      const wrapper = mount(<Button>has loading?</Button>);

      expect(wrapper.find('.c7n-pro-btn-loading').length).toBe(0);
    });

    it('set loading value is false', () => {
      const wrapper = mount(<Button loading={false}>has loading?</Button>);

      expect(wrapper.find('.c7n-pro-btn-loading').length).toBe(0);
    });

    it('set loading value is true', () => {
      const wrapper = mount(<Button loading>has loading?</Button>);

      expect(wrapper.find('.c7n-pro-btn-loading').length).toBe(1);
    });
  })

  describe('Button href', () => {
    it('validator href is accuracy', () => {
      const wrapper = mount(<Button href="https://choerodon.io">https://choerodon.io</Button>);

      expect(wrapper.find('a').props().href).toBe('https://choerodon.io');
    });

    it('not set href and set target will invalid', () => {
      const wrapper = mount(<Button target="_top">https://choerodon.io</Button>);

      // 没有 href 都不会生成 a 标签
      expect(wrapper.find('a').length).toBe(0);
    });

    it('validator target is accuracy', () => {
      const wrapper = mount(<Button href="https://choerodon.io" target="_top">https://choerodon.io</Button>);

      expect(wrapper.find('a').props().href).toBe('https://choerodon.io');
      expect(wrapper.find('a').props().target).toBe('_top');
    });
  })

});
