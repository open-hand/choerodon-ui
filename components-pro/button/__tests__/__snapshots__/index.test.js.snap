// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button button-pro should support link button 1`] = `
<a
  class="c7n-pro-btn c7n-pro-btn-raised c7n-pro-btn-default c7n-pro-btn-wrapper"
  href="https://choerodon.github.io/choerodon-ui/"
  target="_blank"
>
  <span>
    button-pro link
  </span>
  <div
    class="c7n-ripple-wrapper"
  />
</a>
`;

exports[`Button renders Chinese characters correctly 1`] = `
<button
  class="c7n-pro-btn c7n-pro-btn-raised c7n-pro-btn-default c7n-pro-btn-wrapper"
  type="button"
>
  <span>
    按钮
  </span>
  <div
    class="c7n-ripple-wrapper"
  />
</button>
`;

exports[`Button renders Chinese characters correctly 2`] = `
<button
  class="c7n-pro-btn c7n-pro-btn-raised c7n-pro-btn-default c7n-pro-btn-wrapper"
  type="button"
>
  <i
    class="icon icon-search"
  />
  <span>
    按钮
  </span>
  <div
    class="c7n-ripple-wrapper"
  />
</button>
`;

exports[`Button renders Chinese characters correctly 3`] = `
<button
  class="c7n-pro-btn c7n-pro-btn-raised c7n-pro-btn-default c7n-pro-btn-wrapper"
  type="button"
>
  <span>
    <i
      class="icon icon-search"
    />
    按钮
  </span>
  <div
    class="c7n-ripple-wrapper"
  />
</button>
`;

exports[`Button renders components-pro correctly 1`] = `
<button
  class="c7n-pro-btn-wrapper c7n-pro-btn c7n-pro-btn-raised c7n-pro-btn-default"
  type="button"
>
  <span>
     Button-Pro 
  </span>
  <div
    class="c7n-ripple-wrapper"
  />
</button>
`;
